presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -204,7 +207,7 @@ public Rewriter(
             this.partialMergePushdownStrategy = getPartialMergePushdownStrategy(session);
             this.preferStreamingOperators = preferStreamingOperators(session);
             this.partitioningProviderCatalog = getPartitioningProviderCatalog(session);
-            this.hashPartitionCount = getHashPartitionCount(session);
+            this.hashPartitionCount = (physicalResourceOptimizer.isValid()) ? physicalResourceOptimizer.getHashPartitionCount() : getHashPartitionCount(session);
```

Comment:
From what I can gather, its seems that AddExchanges is not actually responsible for setting up the partitioning for PoS. `hashPartitionCount` is only used inside `Partitioning createPartitioning(Collection<VariableReferenceExpression> partitioningColumns)` method. The beginning of this method have a IF block:

```
if (GlobalSystemConnector.NAME.equals(partitioningProviderCatalog)) {
       return Partitioning.create(FIXED_HASH_DISTRIBUTION, ImmutableList.copyOf(partitioningColumns));
}
```

So, if `partitioningProviderCatalog` is equal to `system`, then we just return back `FIXED_HASH_DISTRIBUTION` Partitioning. If you look in `PrestoSparkSettingsRequirements`, there is a `verify()` method which checks for `partitioningProviderCatalog` to be `system` and if it is anything else, it fails the query. 

```
verify(getPartitioningProviderCatalog(session).equals(GlobalSystemConnector.NAME), "partitioning provider other than system is not supported");
``` 

PoS partitioning is configured inside `PrestoSparkQueryExecutionFactory.configureOutputPartitioning()` and we should do probably confine these changes within this class itself.

Reply:
No longer updating AddExchanges. This facilitates moving the resource calculations outside the optimizer. The calculation are done in PrestoSparkPhysicalResourceCalculator in presto/spark. Consequently the PrestoSparkPhysicalResourceCalculator is instantiated in PrestoSparkQueryPlanner.